File: app/src/main/java/io/github/xalrk/nudge/ui/EditReminderScreen.kt
package io.github.xalrk.nudge.ui

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.Delete
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Button
import androidx.compose.material3.DatePicker
import androidx.compose.material3.DatePickerDialog
import androidx.compose.material3.FilterChip
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Scaffold
import androidx.compose.material3.SegmentedButton
import androidx.compose.material3.SegmentedButtonDefaults
import androidx.compose.material3.SingleChoiceSegmentedButtonRow
import androidx.compose.material3.Switch
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.TimePicker
import androidx.compose.material3.TopAppBar
import androidx.compose.material3.rememberDatePickerState
import androidx.compose.material3.rememberTimePickerState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import io.github.xalrk.nudge.data.Kind
import io.github.xalrk.nudge.data.Reminder
import io.github.xalrk.nudge.data.Repeat
import io.github.xalrk.nudge.domain.Recurrence
import java.time.DayOfWeek
import java.time.Instant
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime
import java.time.ZoneId
import java.time.ZoneOffset
import java.time.format.TextStyle
import java.util.Locale

@Composable
fun EditReminderScreen(vm: NudgeViewModel, id: Long, defaultKind: String, onBack: () -> Unit) {
    val reminders by vm.reminders.collectAsStateWithLifecycle()
    val existing = reminders.firstOrNull { it.id == id }

    var loaded by remember { mutableStateOf(id == 0L) }
    var title by remember { mutableStateOf("") }
    var body by remember { mutableStateOf("") }
    var kind by remember { mutableStateOf(runCatching { Kind.valueOf(defaultKind) }.getOrDefault(Kind.SCHEDULED)) }
    var date by remember { mutableStateOf(LocalDate.now()) }
    var time by remember { mutableStateOf(LocalTime.now().plusHours(1).withMinute(0).withSecond(0).withNano(0)) }
    var repeat by remember { mutableStateOf(Repeat.NONE) }
    var intervalText by remember { mutableStateOf("1") }
    var weekdays by remember { mutableStateOf(setOf<DayOfWeek>()) }
    var endDate by remember { mutableStateOf<LocalDate?>(null) }
    var floating by remember { mutableStateOf(true) }
    var zoneId by remember { mutableStateOf(ZoneId.systemDefault().id) }
    var enabled by remember { mutableStateOf(true) }
    var showDate by remember { mutableStateOf(false) }
    var showTime by remember { mutableStateOf(false) }
    var showEnd by remember { mutableStateOf(false) }
    var confirmDelete by remember { mutableStateOf(false) }

    LaunchedEffect(existing) {
        if (!loaded && existing != null) {
            title = existing.title; body = existing.body; kind = existing.kind
            existing.localDateTimeOrNull()?.let { date = it.toLocalDate(); time = it.toLocalTime() }
            repeat = existing.repeat; intervalText = existing.interval.toString()
            weekdays = existing.weekdaySet(); endDate = existing.endDateOrNull()
            floating = existing.floating; zoneId = existing.zoneId ?: ZoneId.systemDefault().id
            enabled = existing.enabled
            loaded = true
        }
    }

    val isNew = id == 0L
    val canSave = title.isNotBlank()

    fun build(): Reminder {
        val base = existing ?: Reminder(title = "", kind = kind)
        return if (kind == Kind.RANDOM) base.copy(
            title = title.trim(), body = body.trim(), kind = Kind.RANDOM,
            localDateTime = null, zoneId = null, repeat = Repeat.NONE, interval = 1, weekdays = 0, endDate = null,
            enabled = enabled, nextAt = if (existing?.isRandom == true) existing.nextAt else null,
        ) else base.copy(
            title = title.trim(), body = body.trim(), kind = Kind.SCHEDULED,
            localDateTime = LocalDateTime.of(date, time).format(Reminder.DT_FORMAT),
            zoneId = if (existing?.zoneId != null && !floating) existing.zoneId else ZoneId.systemDefault().id,
            floating = floating, repeat = repeat, interval = intervalText.toIntOrNull()?.coerceAtLeast(1) ?: 1,
            weekdays = if (repeat == Repeat.WEEKLY) Reminder.maskOf(weekdays) else 0,
            endDate = if (repeat == Repeat.NONE) null else endDate?.toString(),
            enabled = true, nextAt = null, snoozeAt = null,
        )
    }

    Scaffold(
        topBar = {
            TopAppBar(
                title = { Text(if (isNew) "New reminder" else "Edit reminder") },
                navigationIcon = { IconButton(onClick = onBack) { Icon(Icons.AutoMirrored.Filled.ArrowBack, "Back") } },
                actions = {
                    if (!isNew) IconButton(onClick = { confirmDelete = true }) { Icon(Icons.Filled.Delete, "Delete") }
                    TextButton(enabled = canSave, onClick = { vm.save(build()) { onBack() } }) { Text("Save") }
                },
            )
        }
    ) { padding ->
        Column(Modifier.fillMaxSize().padding(padding).verticalScroll(rememberScrollState()).padding(16.dp), verticalArrangement = Arrangement.spacedBy(12.dp)) {
            OutlinedTextField(value = title, onValueChange = { title = it }, label = { Text("Message") }, modifier = Modifier.fillMaxWidth(), singleLine = true)
            OutlinedTextField(value = body, onValueChange = { body = it }, label = { Text("Details (optional)") }, modifier = Modifier.fillMaxWidth(), minLines = 2)

            SingleChoiceSegmentedButtonRow(Modifier.fillMaxWidth()) {
                SegmentedButton(selected = kind == Kind.SCHEDULED, onClick = { kind = Kind.SCHEDULED }, shape = SegmentedButtonDefaults.itemShape(0, 2)) { Text("Scheduled") }
                SegmentedButton(selected = kind == Kind.RANDOM, onClick = { kind = Kind.RANDOM }, shape = SegmentedButtonDefaults.itemShape(1, 2)) { Text("Random") }
            }

            if (kind == Kind.RANDOM) {
                Text("Fires at a random moment during your active hours, at the average frequency set in Settings.",
                    style = MaterialTheme.typography.bodyMedium, color = MaterialTheme.colorScheme.onSurfaceVariant)
                if (!isNew) Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
                    Text("Enabled", Modifier.weight(1f)); Switch(checked = enabled, onCheckedChange = { enabled = it })
                }
            } else {
                Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                    OutlinedButton(onClick = { showDate = true }, Modifier.weight(1f)) { Text(date.format(Fmt.date)) }
                    OutlinedButton(onClick = { showTime = true }) { Text(time.format(Fmt.time)) }
                }

                Text("Repeat", style = MaterialTheme.typography.labelLarge)
                SingleChoiceSegmentedButtonRow(Modifier.fillMaxWidth()) {
                    val labels = listOf("Once", "Day", "Week", "Month", "Year")
                    Repeat.entries.forEachIndexed { i, r ->
                        SegmentedButton(selected = repeat == r, onClick = { repeat = r }, shape = SegmentedButtonDefaults.itemShape(i, Repeat.entries.size)) { Text(labels[i]) }
                    }
                }
                if (repeat != Repeat.NONE) {
                    Row(verticalAlignment = Alignment.CenterVertically) {
                        Text("Every")
                        Spacer(Modifier.width(8.dp))
                        OutlinedTextField(
                            value = intervalText, onValueChange = { intervalText = it.filter(Char::isDigit).take(3) },
                            modifier = Modifier.width(80.dp), singleLine = true,
                            keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number),
                        )
                        Spacer(Modifier.width(8.dp))
                        val n = intervalText.toIntOrNull() ?: 1
                        Text(when (repeat) { Repeat.DAILY -> "day"; Repeat.WEEKLY -> "week"; Repeat.MONTHLY -> "month"; Repeat.YEARLY -> "year"; else -> "" } + if (n == 1) "" else "s")
                    }
                    if (repeat == Repeat.WEEKLY) {
                        Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.spacedBy(4.dp)) {
                            for (d in DayOfWeek.entries) {
                                FilterChip(
                                    selected = d in weekdays || (weekdays.isEmpty() && d == date.dayOfWeek),
                                    onClick = { weekdays = if (d in weekdays) weekdays - d else weekdays + d },
                                    label = { Text(d.getDisplayName(TextStyle.SHORT, Locale.getDefault()).take(2)) },
                                )
                            }
                        }
                    }
                    Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
                        Text("Ends", Modifier.weight(1f))
                        OutlinedButton(onClick = { showEnd = true }) { Text(endDate?.format(Fmt.date) ?: "Never") }
                        if (endDate != null) TextButton(onClick = { endDate = null }) { Text("Clear") }
                    }
                }

                Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
                    Column(Modifier.weight(1f)) {
                        Text("Follow device time zone")
                        Text(
                            if (floating) "Rings at ${time.format(Fmt.time)} local time wherever you are (${ZoneId.systemDefault().id})"
                            else "Pinned to $zoneId; the moment stays fixed when you travel",
                            style = MaterialTheme.typography.bodySmall, color = MaterialTheme.colorScheme.onSurfaceVariant,
                        )
                    }
                    Switch(checked = floating, onCheckedChange = { floating = it })
                }

                val preview = remember(date, time, repeat, intervalText, weekdays, endDate, floating) {
                    val r = build()
                    Recurrence.nextOccurrenceAfter(r, Instant.now())?.withZoneSameInstant(ZoneId.systemDefault())
                }
                Text(
                    if (preview != null) "Next: ${preview.format(Fmt.dayTimeYear)} · ${Recurrence.describe(build())}"
                    else "That time is in the past; nothing will fire.",
                    style = MaterialTheme.typography.bodyMedium,
                    color = if (preview != null) MaterialTheme.colorScheme.primary else MaterialTheme.colorScheme.error,
                )
            }

            Spacer(Modifier.height(8.dp))
            Button(onClick = { vm.save(build()) { onBack() } }, enabled = canSave, modifier = Modifier.fillMaxWidth()) { Text("Save") }
        }
    }

    if (showDate) {
        val state = rememberDatePickerState(initialSelectedDateMillis = date.atStartOfDay(ZoneOffset.UTC).toInstant().toEpochMilli())
        DatePickerDialog(
            onDismissRequest = { showDate = false },
            confirmButton = { TextButton(onClick = {
                state.selectedDateMillis?.let { date = Instant.ofEpochMilli(it).atZone(ZoneOffset.UTC).toLocalDate() }
                showDate = false
            }) { Text("OK") } },
            dismissButton = { TextButton(onClick = { showDate = false }) { Text("Cancel") } },
        ) { DatePicker(state) }
    }
    if (showEnd) {
        val state = rememberDatePickerState(initialSelectedDateMillis = (endDate ?: date.plusMonths(1)).atStartOfDay(ZoneOffset.UTC).toInstant().toEpochMilli())
        DatePickerDialog(
            onDismissRequest = { showEnd = false },
            confirmButton = { TextButton(onClick = {
                state.selectedDateMillis?.let { endDate = Instant.ofEpochMilli(it).atZone(ZoneOffset.UTC).toLocalDate() }
                showEnd = false
            }) { Text("OK") } },
            dismissButton = { TextButton(onClick = { showEnd = false }) { Text("Cancel") } },
        ) { DatePicker(state) }
    }
    if (showTime) {
        val state = rememberTimePickerState(initialHour = time.hour, initialMinute = time.minute)
        AlertDialog(
            onDismissRequest = { showTime = false },
            confirmButton = { TextButton(onClick = { time = LocalTime.of(state.hour, state.minute); showTime = false }) { Text("OK") } },
            dismissButton = { TextButton(onClick = { showTime = false }) { Text("Cancel") } },
            text = { TimePicker(state) },
        )
    }
    if (confirmDelete) AlertDialog(
        onDismissRequest = { confirmDelete = false },
        title = { Text("Delete reminder?") },
        text = { Text("\"$title\" will be removed.") },
        confirmButton = { TextButton(onClick = { confirmDelete = false; vm.delete(id); onBack() }) { Text("Delete") } },
        dismissButton = { TextButton(onClick = { confirmDelete = false }) { Text("Cancel") } },
    )
}

File: ios/Nudge/Sources/UI/EditReminderScreen.swift
import SwiftUI
import NudgeCore

/// Create or edit one reminder. Presented full screen; Cancel asks before discarding changes.
struct EditReminderScreen: View {
    @EnvironmentObject private var engine: ReminderEngine
    @EnvironmentObject private var store: Store
    @EnvironmentObject private var settings: SettingsStore
    @EnvironmentObject private var router: Router
    @Environment(\.theme) private var theme

    let request: EditRequest

    @State private var loaded = false
    @State private var title = ""
    @State private var body_ = ""
    @State private var kind: Kind = .scheduled
    @State private var date = LocalDate.today
    @State private var time = LocalTime(0, 0)
    @State private var repeatRule: Repeat = .none
    @State private var intervalText = "1"
    @State private var weekdays = Set<DayOfWeek>()
    @State private var endDate: LocalDate? = nil
    @State private var floating = true
    @State private var zoneId = TimeZone.current.identifier
    @State private var enabled = true
    @State private var color: Int32? = nil
    @State private var sound = true
    @State private var customRate = false
    @State private var customDays = false
    @State private var randomDays = Set<DayOfWeek>()
    @State private var rateSlider = SettingsMath.millisToSlider(SettingsMath.defaultMeanMillis)
    @State private var initial: Reminder? = nil
    @State private var confirmDelete = false
    @State private var confirmDiscard = false
    @State private var askScope: String? = nil // "save" or "delete"

    private var existing: Reminder? { store.state.reminders.first { $0.id == request.id } }
    private var isNew: Bool { request.id == 0 }
    private var canSave: Bool { !title.isBlank }
    private var isSeries: Bool { existing != nil && existing!.isScheduled && existing!.repeatRule != .none && request.occurrence != nil }
    private var s: SettingsSnapshot { settings.snapshot }

    private func load() {
        guard !loaded else { return }
        kind = request.kind
        date = request.date ?? LocalDate.today
        let now = LocalDateTime.now
        time = LocalTime((now.time.hour + 1) % 24, 0)
        if now.time.hour == 23 { date = date.plusDays(1) }
        rateSlider = SettingsMath.millisToSlider(s.meanIntervalMillis)
        randomDays = s.activeDaySet()
        if let e = existing {
            title = e.title; body_ = e.body; kind = e.kind
            if let ldt = e.localDateTimeOrNil() { date = request.occurrence ?? ldt.date; time = ldt.time }
            repeatRule = e.repeatRule; intervalText = String(e.interval)
            weekdays = e.weekdaySet(); endDate = e.endDateOrNil()
            floating = e.floating; zoneId = e.zoneId ?? TimeZone.current.identifier
            enabled = e.enabled; color = e.color; sound = e.sound
            customRate = e.meanOverrideMillis != nil
            customDays = e.isRandom && e.weekdays != 0
            randomDays = e.randomDaysOrNil() ?? s.activeDaySet()
            rateSlider = SettingsMath.millisToSlider(e.meanOverrideMillis ?? s.meanIntervalMillis)
        }
        loaded = true
        initial = signature()
    }

    private func build() -> Reminder {
        var r = existing ?? Reminder(title: "", kind: kind)
        r.title = title.trimmingCharacters(in: .whitespacesAndNewlines)
        r.body = body_.trimmingCharacters(in: .whitespacesAndNewlines)
        r.color = color; r.sound = sound; r.vibrate = true
        if kind == .random {
            let mean: Int64? = customRate ? SettingsMath.sliderToMillis(rateSlider) : nil
            let mask = (customDays && !randomDays.isEmpty) ? Reminder.maskOf(randomDays) : 0
            r.kind = .random; r.localDateTime = nil; r.zoneId = nil; r.repeatRule = .none; r.interval = 1; r.endDate = nil
            r.weekdays = mask; r.excludedDates = ""; r.enabled = enabled; r.meanOverrideMillis = mean
            // A changed rate or day set needs a fresh roll; otherwise keep the pending time.
            r.nextAt = (existing?.isRandom == true && existing!.meanOverrideMillis == mean && existing!.weekdays == mask) ? existing!.nextAt : nil
        } else {
            r.meanOverrideMillis = nil
            r.kind = .scheduled
            r.localDateTime = date.atTime(time).description
            r.zoneId = (existing?.zoneId != nil && !floating) ? existing!.zoneId : TimeZone.current.identifier
            r.floating = floating; r.repeatRule = repeatRule; r.interval = max(Int(intervalText) ?? 1, 1)
            r.weekdays = repeatRule == .weekly ? Reminder.maskOf(weekdays) : 0
            r.endDate = repeatRule == .none ? nil : endDate?.description
            r.excludedDates = repeatRule == .none ? "" : (existing?.excludedDates ?? "")
            r.enabled = true; r.nextAt = nil; r.snoozeAt = nil
        }
        return r
    }

    /// The editable fields only, so two builds compare equal when nothing the user can change differs.
    private func signature() -> Reminder {
        var r = build()
        r.id = 0; r.nextAt = nil; r.snoozeAt = nil; r.lastFiredAt = nil; r.createdAt = 0; r.dedupeKey = ""; r.zoneId = nil
        return r
    }

    private var dirty: Bool { loaded && initial != nil && signature() != initial }

    private func leave() { if dirty { confirmDiscard = true } else { router.edit = nil } }

    private func save() {
        if isSeries && kind == .scheduled && repeatRule != .none { askScope = "save" }
        else if engine.save(build()) { router.edit = nil }
    }

    var body: some View {
        NavigationStack {
            Screen {
                ScrollView {
                    VStack(alignment: .leading, spacing: 12) {
                        TextField("Message", text: $title).textFieldStyle(.roundedBorder)
                        TextField("Details (optional)", text: $body_, axis: .vertical).lineLimit(2...6).textFieldStyle(.roundedBorder)
                        Segmented(options: [(Kind.scheduled, "Scheduled"), (Kind.random, "Random")], selection: $kind)

                        if kind == .random { randomSection } else { scheduledSection }

                        Text("Notification").font(.subheadline.weight(.semibold)).padding(.top, 4)
                        Toggle("Sound", isOn: $sound)
                        Hint("Silent reminders still show a banner. iOS decides vibration from your sound settings.")
                        Text("Color").font(.body)
                        Hint("Shown on the calendar. \"A\" follows the app accent (its complementary color).")
                        SwatchRow(current: color, customColors: s.customColors, autoColor: Colors.complementary(s.accentColor),
                                  onPick: { color = $0 }, onAddCustom: { engine.addCustomColor($0) },
                                  onRemoveCustom: { removed in engine.removeCustomColor(removed); if color == removed { color = nil } })
                        Spacer(minLength: 24)
                    }
                    .padding(16)
                }
            }
            .navigationTitle(isNew ? "New Reminder" : "Edit Reminder")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) { Button("Cancel") { leave() } }
                ToolbarItemGroup(placement: .confirmationAction) {
                    if !isNew {
                        Button { if isSeries { askScope = "delete" } else { confirmDelete = true } } label: { Image(systemName: "trash") }
                            .accessibilityLabel("Delete")
                    }
                    Button("Save") { save() }.disabled(!canSave).fontWeight(.semibold)
                }
            }
        }
        .interactiveDismissDisabled(dirty)
        .onAppear(perform: load)
        .alert("Discard changes?", isPresented: $confirmDiscard) {
            Button("Discard", role: .destructive) { router.edit = nil }
            Button("Keep editing", role: .cancel) { }
        } message: { Text("You have unsaved changes to this reminder.") }
        .alert("Delete reminder?", isPresented: $confirmDelete) {
            Button("Delete", role: .destructive) { engine.delete(request.id); router.edit = nil }
            Button("Cancel", role: .cancel) { }
        } message: { Text("\"\(title)\" will be removed.") }
        .confirmationDialog(askScope == "delete" ? "Delete repeating reminder" : "Change repeating reminder",
                            isPresented: Binding(get: { askScope != nil }, set: { if !$0 { askScope = nil } }), titleVisibility: .visible) {
            if let occ = request.occurrence, let orig = existing {
                let deleting = askScope == "delete"
                Button("Only this one (\(Fmt.date(occ)))") { runScope(.this, deleting: deleting, orig: orig, occ: occ) }
                Button("This and following") { runScope(.following, deleting: deleting, orig: orig, occ: occ) }
                Button("All occurrences", role: deleting ? ButtonRole.destructive : nil) { runScope(.all, deleting: deleting, orig: orig, occ: occ) }
                Button("Cancel", role: .cancel) { askScope = nil }
            }
        } message: { Text("This reminder repeats. Apply to:") }
    }

    private func runScope(_ scope: SeriesScope, deleting: Bool, orig: Reminder, occ: LocalDate) {
        askScope = nil
        if deleting { engine.deleteFromSeries(original: orig, occDate: occ, scope: scope); router.edit = nil }
        else if engine.editSeries(original: orig, edited: build(), occDate: occ, scope: scope) { router.edit = nil }
    }

    // MARK: sections

    @ViewBuilder private var randomSection: some View {
        Hint("Fires at a random moment during your active hours.")
        LabeledToggle(title: "Custom frequency",
                      subtitle: customRate ? SettingsMath.describeInterval(SettingsMath.sliderToMillis(rateSlider)).capitalizedFirst
                                           : "Uses the Settings default (\(SettingsMath.describeInterval(s.meanIntervalMillis)))",
                      subtitleAccent: customRate, isOn: $customRate)
        if customRate { Slider(value: $rateSlider, in: 0...1) }
        LabeledToggle(title: "Custom days", subtitle: customDays ? "Only on the selected days" : "Uses the Settings active days", subtitleAccent: customDays,
                      isOn: Binding(get: { customDays }, set: { on in customDays = on; if on && randomDays.isEmpty { randomDays = s.activeDaySet() } }))
        if customDays {
            DayCircles(selected: randomDays) { d in
                let next = randomDays.contains(d) ? randomDays.subtracting([d]) : randomDays.union([d])
                if !next.isEmpty { randomDays = next }
            }
        }
        if !isNew { Toggle("Enabled", isOn: $enabled) }
    }

    @ViewBuilder private var scheduledSection: some View {
        HStack {
            DatePicker("Date", selection: Binding(get: { date.foundationDate }, set: { date = LocalDate(foundationDate: $0) }), displayedComponents: .date).labelsHidden()
            Spacer()
            DatePicker("Time", selection: Binding(get: { time.foundationDate }, set: { time = LocalTime(foundationDate: $0) }), displayedComponents: .hourAndMinute).labelsHidden()
        }
        // Quick picks for the common "remind me soon" cases.
        let now = LocalDateTime.now
        let tonight: LocalDateTime = { let t = now.date.atTime(20, 0); return t.isAfter(now) ? t : t.plusDays(1) }()
        VStack(alignment: .leading, spacing: 6) {
            HStack(spacing: 8) {
                Chip("In 15 min") { pick(now.plusMinutes(15)) }
                Chip("In 1 hour") { pick(now.plusHours(1)) }
                Chip("Tonight 8 pm") { pick(tonight) }
            }
            HStack(spacing: 8) {
                Chip("Tomorrow 9 am") { pick(now.date.plusDays(1).atTime(9, 0)) }
                Chip("Next week") { pick(now.date.plusWeeks(1).atTime(time)) }
            }
        }

        Text("Repeat").font(.subheadline.weight(.semibold))
        Segmented(options: [(Repeat.none, "Once"), (.daily, "Day"), (.weekly, "Week"), (.monthly, "Month"), (.yearly, "Year")], selection: $repeatRule)
        if repeatRule != .none {
            HStack(spacing: 8) {
                Text("Every")
                TextField("1", text: Binding(get: { intervalText }, set: { intervalText = String($0.filter { $0.isNumber }.prefix(3)) }))
                    .keyboardType(.numberPad).textFieldStyle(.roundedBorder).frame(width: 70)
                let n = Int(intervalText) ?? 1
                Text(unitName + (n == 1 ? "" : "s"))
            }
            if repeatRule == .weekly {
                let effective: Set<DayOfWeek> = weekdays.isEmpty ? [date.dayOfWeek] : weekdays
                DayCircles(selected: effective) { d in
                    if weekdays.contains(d) { weekdays.remove(d) } else { weekdays.insert(d) }
                }
            }
            HStack {
                Text("Ends")
                Spacer()
                if let end = endDate {
                    DatePicker("Ends", selection: Binding(get: { end.foundationDate }, set: { endDate = LocalDate(foundationDate: $0) }), displayedComponents: .date).labelsHidden()
                    Button("Clear") { endDate = nil }
                } else {
                    OutlineButton(title: "Never") { endDate = date.plusMonths(1) }
                }
            }
        }

        LabeledToggle(title: "Follow device time zone",
                      subtitle: floating ? "Rings at \(Fmt.time(time)) local time wherever you are (\(TimeZone.current.identifier))"
                                         : "Pinned to \(zoneId); the moment stays fixed when you travel",
                      isOn: $floating)

        let preview = Recurrence.nextOccurrenceAfter(build(), Date())?.withZone(.device)
        Text(preview != nil ? "Next: \(Fmt.dayTimeYear(preview!)) · \(Recurrence.describe(build()))" : "That time is in the past; nothing will fire.")
            .font(.subheadline).foregroundColor(preview != nil ? theme.primary : theme.error)
    }

    private var unitName: String {
        switch repeatRule { case .daily: return "day"; case .weekly: return "week"; case .monthly: return "month"; case .yearly: return "year"; case .none: return "" }
    }

    private func pick(_ dt: LocalDateTime) { date = dt.date; time = LocalTime(dt.time.hour, dt.time.minute) }
}

private struct Chip: View {
    @Environment(\.theme) private var theme
    let label: String
    let action: () -> Void
    init(_ label: String, action: @escaping () -> Void) { self.label = label; self.action = action }
    var body: some View {
        Button(action: action) {
            Text(label).font(.footnote).foregroundColor(theme.onSurface)
                .padding(.horizontal, 12).padding(.vertical, 7)
                .overlay(RoundedRectangle(cornerRadius: 8).strokeBorder(theme.outline, lineWidth: 1))
        }
        .buttonStyle(.plain)
    }
}

extension String {
    var capitalizedFirst: String { prefix(1).uppercased() + dropFirst() }
}
